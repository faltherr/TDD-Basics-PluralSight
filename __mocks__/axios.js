const axiosMock = jest.genMockFromModule('axios');

// Specify what the mock response should be and the data we want to test against
// Should have the same shape as the data that's current.ly inside of the component to test

let mockResponse = {
    data: {
        "shops": [
            {
                "location": "test Location",
                "address": "test Address"
            }
    ]}
}

//Creat a mock implementation of the GET method from axios
axiosMock.get.mockImplementation(req);

function req(){
    return new Promise(function(resolve){
        axiosMock.delayTimer = setTimeout(function(){
            resolve(mockResponse)
        }, 100);
    })
}

module.exports = axiosMock;