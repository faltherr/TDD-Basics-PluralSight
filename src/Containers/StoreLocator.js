import React, {Component} from 'react'
import Header from '../Components/Header'
import Button from '../Components/Button'
import Map from '../Components/Map'
import mapChooser from '../mapChooser'
import axios from 'axios';

class StoreLocator extends Component{
    constructor(props){
        super(props);
        this.state = {
            currentMap : "none.png",
            shops: []
        }
        this.shops = [
            {
            'location': 'Portland',
            'address': '123 Portland Dr'
            },
            {
            'location': 'Astoria',
            'address': '456 Astoria Dr'
            },
            {            
            'location': '',
            'address': ''
            }
        ]
    this.chooseMap = this.chooseMap.bind(this)
    }

    async componentDidMount(){
        let response = await axios.get('http://localhost:3000/data/shops.json')
        this.setState({
            shops: response.data.shops
        })
    }

    chooseMap(e){
        // Value is defined in the button component
        this.setState({currentMap:mapChooser(e.target.value)})
    }

    render(){
        let storeButtons = this.state.shops.map((shop,id) => {
            return(
                <Button handleClick={this.chooseMap} key={id} location={shop.location}/>
            )
        })
        return(
            <div>
                <Header/>
                <div>
                    {storeButtons}
                </div>
                <Map imageName={this.state.currentMap} location={this.props.location}/>
            </div>
        )
    }
}

export default StoreLocator