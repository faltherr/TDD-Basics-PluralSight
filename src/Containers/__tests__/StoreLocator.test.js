import React from "react";
import { shallow } from "enzyme";
import StoreLocator from "../StoreLocator";
import axios from "axios";
import renderer from 'react-test-renderer';


describe("StoreLocator", function() {
  let mountedStoreLocator;
  //Use the same store locator before each of our tests
  beforeEach(() => {
    mountedStoreLocator = shallow(<StoreLocator />);
  });

  it("Renders the correct snapshot", ()=>{
    const tree = renderer.create(<StoreLocator/>).toJSON()
    expect(tree).toMatchSnapshot();
  });

  it("Calls axios.get in componentDidMount", () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        expect(axios.get).toHaveBeenCalled();
      });
  });

  it("Calls axios.get with the correct url", () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        expect(axios.get).toHaveBeenCalledWith(
          "http://localhost:3000/data/shops.json"
        );
      });
  });

  it("updates state with data from the api", () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        //Expect the value to be equal to our mock return value in the mock axios function
        expect(mountedStoreLocator.state()).toHaveProperty("shops", [
            {
              location: "test Location",
              address: "test Address"
            }
          ]);
      });
  });

  it("renders without crashing", () => {
    let mountedStoreLocator = shallow(<StoreLocator />);
  });

  it("renders a header", () => {
    const headers = mountedStoreLocator.find("Header");
    expect(headers.length).toBe(1);
  });

  it("renders a map", () => {
    const map = mountedStoreLocator.find("Map");
    expect(map.length).toBe(1);
  });
});

describe("chooseMap", () => {
  it("updates this.state.currentMap based on the location passed into it", () => {
    let mountedStoreLocator = shallow(<StoreLocator />);
    // Simulate the event comming from the button component
    let mockEvent = { target: { value: "testland" } };
    mountedStoreLocator.instance().chooseMap(mockEvent);
    expect(mountedStoreLocator.instance().state.currentMap).toBe(
      "testland.png"
    );
  });
});
