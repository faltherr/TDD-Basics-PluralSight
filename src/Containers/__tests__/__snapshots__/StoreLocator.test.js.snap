// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreLocator Renders the correct snapshot 1`] = `
<div>
  <div
    className="Header"
  >
    <img
      alt="my coffee shop"
      src="images/logo.png"
    />
  </div>
  <div />
  <div
    className="MapBox"
  >
    <img
      src="images/none.png"
    />
  </div>
</div>
`;
